### ISSUE TITLE
Can we add encryption to the database?

### ISSUE THREAD

**Issue by AlexTrotsenko** [Kodein-Framework/Kodein-DB#4]
After checking the [talk at KotlinConf](https://youtu.be/2mpzbFxUkAE) I found this project interesting to try. 

But in order to use it in production I would need to have ability to have the data _encrypted_ when they are stored. 

I assume, that right now you don't have any kind of encryption implemented (e.g. via [EncryptedFile](https://developer.android.com/reference/androidx/security/crypto/EncryptedFile), etc.) Can you advice if it's indeed the case?

If so -  can you advise if custom encryption could be defined already now or some changes would be required? On [this slide](https://youtu.be/2mpzbFxUkAE?t=645) I see, that there is an option to provide "custom middlewares"  for transforming data. Do you know if it's suitable for this purpose?

Basically we need to have an api which transforms un-encrypted bytes to encrypted bytes when we `.put()` and vise versa on `.get()`.

As example you can check this [Encryption](https://github.com/nolia/Noodle/blob/master/noodle/src/main/java/com/noodle/storage/Encryption.java) interface.

**Comment by SalomonBrys** [comment]
Yes, Data Middleware is the way to go.
I don't think it would be very hard to implement, I could provide a sample.
However, you need to understand that only the object serialized bytes would be encrypted.

Here is a list of things that **cannot be encrypted**, because their part of LevelDB keys:
- Primary keys (not problem if it is a random `UUID`, but definitely problematic if it is a String with semantic content).
- Index **values**, since the value that is indexed needs to be ordered with other values.
- Object's list of index

I suppose you could also make your data middleware ensure that no indices are declared on an encrypted object, to make sure that you are not leaking data.
Alternatively, you could also encrypt index values, at the cost of loosing index value ordering (iterating over an index would provide object in a random order).

Thoughts?

**Comment by Talgat777** [comment]
+1

**Comment by ln-12** [comment]
I would also like to see encryption. For SQLDelight, SQLCipher can be used: https://touchlab.co/sqlcipher-kotlin-multiplatform-mobile/

**Comment by gigonzalezs** [comment]
+1

**Comment by gigonzalezs** [comment]
I would be glad to collaborate with this feature

**Comment by SalomonBrys** [comment]
Hey guys, thans for the interest in Kodein-DB.
I am currently working on this, It should be available on the next version of Kodein-DB.
What cipher are you expecting to be available by default ?
I am guessing AES-CBC for document content & HMAC-SHA256 for index values.

**Comment by ln-12** [comment]
I see nothing speaking against those 👍

**Comment by SalomonBrys** [comment]
Hi guys.
Encryption is available in Kodein `0.8.0-beta`.
Documentation is here: https://docs.kodein.org/kodein-db/0.8/core/encryption.html

I hope it fits your needs.
I advise reading the entire document, and especially the "Index metadata & relationships" section.

**Comment by AChep** [comment]
@SalomonBrys what artifact do I need to depend on to get that?

**Comment by Ynnck123** [comment]
I also cannot find the required dependencies for the "_Encryption middleware_", mentioned in the [docmentations](https://docs.kodein.org/kodein-db/0.8/core/encryption.html)
